scalac `-javabootclasspath` option not working as expected

I have a source code that compiles only from a certain JDK version ([8u251+, released on April 2020](https://www.oracle.com/java/technologies/javase/8u251-relnotes.html)). I have attached the source code at the bottom in two formats -- one is a java file and the second is scala.

`getParams` is missing from `java.security.interfaces.RSAKey` in JDK versions earlier than 8u251 (such as 8u241), and thus the compilation should fail in these versions.

I am able to tell `javac` to compile the code with 8u241 bootclasspath, which will fail the compilation as expected:
```
$ java -version
openjdk version "11.0.9" 2020-10-20
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.9+11)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.9+11, mixed mode)

$ javac sources/file.java

$ javac -source 8 -target 8 -bootclasspath /Library/Java/JavaVirtualMachines/jdk1.8.0_241.jdk/Contents/Home/jre/lib/rt.jar sources/file.java

sources/file.java:14: error: cannot find symbol
        new RSAPublicKeySpec(rsa.getModulus(), rsa.getPublicExponent(), rsa.getParams());
                                                                           ^
  symbol:   method getParams()
  location: variable rsa of type RSAPublicKey
1 error
```

However running the same with scalac (installed with Homebrew) will yield a strange error:
```
$ /usr/local/opt/scala@2.12/bin/scalac -version
Scala compiler version 2.12.12 -- Copyright 2002-2020, LAMP/EPFL and Lightbend, Inc.

$ /usr/local/opt/scala@2.12/bin/scalac -release 8 -target:jvm-1.8 -javabootclasspath /Library/Java/JavaVirtualMachines/jdk1.8.0_241.jdk/Contents/Home/jre/lib/rt.jar sources/file.java

error: error while loading Object, Missing dependency 'object scala.native in compiler mirror', required by /876/java/lang/Object.sig
error: scala.reflect.internal.MissingRequirementError: object scala in compiler mirror not found.
	at scala.reflect.internal.MissingRequirementError$.signal(MissingRequirementError.scala:24)
	at scala.reflect.internal.MissingRequirementError$.notFound(MissingRequirementError.scala:25)
	at scala.reflect.internal.Mirrors$RootsBase.$anonfun$getModuleOrClass$5(Mirrors.scala:61)
	at scala.reflect.internal.Mirrors$RootsBase.getPackage(Mirrors.scala:61)
	at scala.reflect.internal.Definitions$DefinitionsClass.ScalaPackage$lzycompute(Definitions.scala:196)
	at scala.reflect.internal.Definitions$DefinitionsClass.ScalaPackage(Definitions.scala:196)
	at scala.reflect.internal.Definitions$DefinitionsClass.ScalaPackageClass$lzycompute(Definitions.scala:197)
	at scala.reflect.internal.Definitions$DefinitionsClass.ScalaPackageClass(Definitions.scala:197)
	at scala.reflect.internal.Definitions$DefinitionsClass.AnyClass$lzycompute(Definitions.scala:292)
	at scala.reflect.internal.Definitions$DefinitionsClass.AnyClass(Definitions.scala:292)
	at scala.tools.nsc.symtab.classfile.ClassfileParser$ClassTypeCompleter.complete(ClassfileParser.scala:1274)
	at scala.reflect.internal.Symbols$Symbol.completeInfo(Symbols.scala:1542)
	at scala.reflect.internal.Symbols$Symbol.info(Symbols.scala:1514)
	at scala.reflect.internal.Symbols$Symbol.initialize(Symbols.scala:1698)
	at scala.reflect.internal.Definitions$DefinitionsClass.init(Definitions.scala:1503)
	at scala.tools.nsc.Global$Run.<init>(Global.scala:1207)
	at scala.tools.nsc.Driver.doCompile(Driver.scala:46)
	at scala.tools.nsc.MainClass.doCompile(Main.scala:32)
	at scala.tools.nsc.Driver.process(Driver.scala:67)
	at scala.tools.nsc.Driver.main(Driver.scala:80)
	at scala.tools.nsc.Main.main(Main.scala)
```

Note that running with `JAVA_HOME` set to the older version will not fail the comilation:
```
$ JAVA_HOME=/Library/Java/JavaVirtualMachines/jdk1.8.0_241.jdk/Contents/Home /usr/local/opt/scala@2.12/bin/scalac  sources/file.java
```

The only thing that does seem to work somehow is trying to compile the Scala file with `JAVA_HOME` set to the older version:

```
$ JAVA_HOME=/Library/Java/JavaVirtualMachines/jdk1.8.0_241.jdk/Contents/Home /usr/local/opt/scala@2.12/bin/scalac  sources/file.scala
sources/file.scala:19: error: too many arguments (3) for constructor RSAPublicKeySpec: (x$1: java.math.BigInteger, x$2: java.math.BigInteger)java.security.spec.RSAPublicKeySpec
            new RSAPublicKeySpec(rsa.getModulus, rsa.getPublicExponent, rsa.getParams)
                                                                            ^
one error found
```

---

The source files:

```scala
// scala
import java.math.BigInteger
import java.security.KeyPairGenerator
import java.security.interfaces.{RSAKey, RSAPublicKey}
import java.security.spec.{AlgorithmParameterSpec, RSAPublicKeySpec}
import javax.crypto.KeyGenerator

object Rsa extends App {
    val kg = KeyPairGenerator.getInstance("RSA")
    val kp = kg.generateKeyPair()
    kp.getPublic match {
        case rsa: RSAPublicKey =>
            new RSAPublicKeySpec(rsa.getModulus, rsa.getPublicExponent, rsa.getParams)
    }
}
```

```java
// java
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.interfaces.RSAPublicKey;
import java.security.spec.RSAPublicKeySpec;

public class Rsa {
    public static void main(String[] args) throws Exception {
        KeyPairGenerator kg = KeyPairGenerator.getInstance("RSA");
        KeyPair keyPair = kg.generateKeyPair();
        RSAPublicKey rsa = (RSAPublicKey) keyPair.getPublic();
        new RSAPublicKeySpec(rsa.getModulus(), rsa.getPublicExponent(), rsa.getParams());
    }
}
```

My example makes it clear what scalac will fail in a Java source, namely, only what Scala needs for typechecking.

The case of direct usage is:
```
➜  snips cat walker.scala

object Walker {
  def walk(walker: StackWalker): Unit = ()
}
object Test extends App {
  println {
    Walker.walk(null)
    //JWalker.walk(null)
    //JWalker.walk2()
    "hello, world!"
  }
}
➜  snips scalac -release 8 walker.scala
walker.scala:5: error: not found: type StackWalker
  def walk(walker: StackWalker): Unit = ()
                   ^
one error found
➜  snips scalac -release 9 walker.scala
➜  snips
```
Verifying arbitrary binary dependencies is not what the feature means.

Just to show that javac does not do that:
```
➜  snips javac --release 8 JWalker.java
JWalker.java:3: error: cannot find symbol
        public static void walk(StackWalker w) {
                                ^
  symbol:   class StackWalker
  location: class JWalker
JWalker.java:6: error: cannot find symbol
                StackWalker w = null;
                ^
  symbol:   class StackWalker
  location: class JWalker
2 errors
➜  snips javac --release 9 JWalker.java
➜  snips javac --release 8 JClient.java
➜  snips
```
`javac --release 8` compiles source that uses a class that does not compile under `--release 8`.

Yes, I closed the issue because I thought it was about how scalac processes java sources.

Please bear with me as I reproduce your issue. I'm on Ubuntu instead of MacOS at the moment.

I'm using the old java 8 I happen to have installed locally. The test file is the Scala source above.

```
➜  snips sdk use  java 8.0.232-zulu

Using java version 8.0.232-zulu in this shell.
➜  snips scalac -d /tmp t12257.scala
t12257.scala:13: error: too many arguments (found 3, expected 2) for constructor RSAPublicKeySpec: (x$1: java.math.BigInteger, x$2: java.math.BigInteger): java.security.spec.RSAPublicKeySpec
      new RSAPublicKeySpec(rsa.getModulus, rsa.getPublicExponent, rsa.getParams)
                                                                      ^
1 error
➜  snips type javac
javac is /home/amarki/.sdkman/candidates/java/8.0.232-zulu/bin/javac
➜  snips ls /home/amarki/.sdkman/candidates/java/8.0.232-zulu/jre/lib/rt.jar
/home/amarki/.sdkman/candidates/java/8.0.232-zulu/jre/lib/rt.jar
➜  snips scalac -d /tmp -nobootcp -javabootclasspath /home/amarki/.sdkman/candidates/java/8.0.232-zulu/jre/lib/rt.jar t12257.scala
t12257.scala:6: error: object crypto is not a member of package javax
import javax.crypto.KeyGenerator
             ^
t12257.scala:13: error: too many arguments (found 3, expected 2) for constructor RSAPublicKeySpec: (x$1: java.math.BigInteger, x$2: java.math.BigInteger): java.security.spec.RSAPublicKeySpec
      new RSAPublicKeySpec(rsa.getModulus, rsa.getPublicExponent, rsa.getParams)
                                                                      ^
2 errors
➜  snips
```
I get the same result with scala 2.12.12.

The constructor is also since java 11, same as getParams:

https://docs.oracle.com/en/java/javase/15/docs/api/java.base/java/security/spec/RSAPublicKeySpec.html#%3Cinit%3E(java.math.BigInteger,java.math.BigInteger,java.security.spec.AlgorithmParameterSpec)

Using just the getParams, I still see:

```
t12257b.scala:12: error: value getParams is not a member of java.security.interfaces.RSAPublicKey
    case rsa: RSAPublicKey => rsa.getParams
                                  ^
```
Presumably something else is going on in your environment.

BTW, I also tried unsetting SCALA_HOME and JAVA_HOME.

It looks like java on PATH overrides any args to `scalac`. This is not reflected in `-Vclasspath`.

Scala 3 has the same behavior.

```
➜  snips sdk use  java 8.0.232-zulu

Using java version 8.0.232-zulu in this shell.
➜  snips scalac -d /tmp -nobootcp -javabootclasspath /home/amarki/.sdkman/candidates/java/8.0.232-zulu/jre/lib/rt.jar t12257.scala
t12257.scala:6: error: object crypto is not a member of package javax
import javax.crypto.KeyGenerator
             ^
t12257.scala:13: error: too many arguments (found 3, expected 2) for constructor RSAPublicKeySpec: (x$1: java.math.BigInteger, x$2: java.math.BigInteger): java.security.spec.RSAPublicKeySpec
      new RSAPublicKeySpec(rsa.getModulus, rsa.getPublicExponent, rsa.getParams)
                                                                      ^
2 errors
➜  snips scalac -d /tmp -nobootcp t12257.scala
t12257.scala:13: error: too many arguments (found 3, expected 2) for constructor RSAPublicKeySpec: (x$1: java.math.BigInteger, x$2: java.math.BigInteger): java.security.spec.RSAPublicKeySpec
      new RSAPublicKeySpec(rsa.getModulus, rsa.getPublicExponent, rsa.getParams)
                                                                      ^
1 error
➜  snips sdk use java 11.0.9.hs-adpt

Using java version 11.0.9.hs-adpt in this shell.
➜  snips scalac -d /tmp -nobootcp -javabootclasspath /home/amarki/.sdkman/candidates/java/8.0.232-zulu/jre/lib/rt.jar t12257.scala
➜  snips scalac -d /tmp -nobootcp t12257.scala
➜  snips
```

--- a/src/compiler/scala/tools/util/PathResolver.scala
+++ b/src/compiler/scala/tools/util/PathResolver.scala
@@ -16,12 +16,13 @@ package util
 
 import java.net.URL
 
+import scala.reflect.io.{Directory, File, Path}
 import scala.tools.reflect.WrappedProperties.AccessControl
 import scala.tools.nsc.{CloseableRegistry, Settings}
-import scala.tools.nsc.util.ClassPath
-import scala.reflect.io.{Directory, File, Path}
-import PartialFunction.condOpt
 import scala.tools.nsc.classpath._
+import scala.tools.nsc.util.ClassPath
+import scala.util.Properties.isJavaAtLeast
+import PartialFunction.condOpt
 
 // Loosely based on the draft specification at:
 // https://wiki.scala-lang.org/display/SIW/Classpath
@@ -211,22 +212,21 @@ final class PathResolver(settings: Settings, closeableRegistry: CloseableRegistr
 
   private val classPathFactory = new ClassPathFactory(settings, closeableRegistry)
 
-  import PathResolver.{ AsLines, Defaults, ppcp }
+  import PathResolver.{AsLines, Defaults, ppcp}
 
-  private def cmdLineOrElse(name: String, alt: String) = {
-    (commandLineFor(name) match {
-      case Some("") => None
-      case x        => x
-    }) getOrElse alt
-  }
+  private def cmdLineOrElse(name: String, alt: String) =
+    commandLineFor(name) match {
+      case Some("") | None => alt
+      case Some(x)         => x
+    }
 
   private def commandLineFor(s: String): Option[String] = condOpt(s) {
-    case "javabootclasspath"  => settings.javabootclasspath.value
-    case "javaextdirs"        => settings.javaextdirs.value
-    case "bootclasspath"      => settings.bootclasspath.value
-    case "extdirs"            => settings.extdirs.value
-    case "classpath" | "cp"   => settings.classpath.value
-    case "sourcepath"         => settings.sourcepath.value
+    case "javabootclasspath" => settings.javabootclasspath.value
+    case "javaextdirs"       => settings.javaextdirs.value
+    case "bootclasspath"     => settings.bootclasspath.value
+    case "extdirs"           => settings.extdirs.value
+    case "classpath" | "cp"  => settings.classpath.value
+    case "sourcepath"        => settings.sourcepath.value
   }
 
   /** Calculated values based on any given command line options, falling back on
@@ -258,8 +258,10 @@ final class PathResolver(settings: Settings, closeableRegistry: CloseableRegistr
 
     // Assemble the elements!
     def basis = List[Iterable[ClassPath]](
-      jrt,                                          // 0. The Java 9+ classpath (backed by the ct.sym or jrt:/ virtual system, if available)
-      classesInPath(javaBootClassPath),             // 1. The Java bootstrap class path.
+      if (!settings.javabootclasspath.isSetByUser && isJavaAtLeast(9))
+        jrt                                         // 0. The Java 9+ classpath (backed by the ct.sym or jrt:/ virtual system, if available)
+      else
+        classesInPath(javaBootClassPath),           // 1. The Java bootstrap class path.
       contentsOfDirsInPath(javaExtDirs),            // 2. The Java extension class path.
       classesInExpandedPath(javaUserClassPath),     // 3. The Java application class path.
       classesInPath(scalaBootClassPath),            // 4. The Scala boot class path.

